File: server/internal/models/report.go
package models

import (
	"time"
)

// ReportStatus 举报状态
type ReportStatus string

const (
	ReportStatusPending    ReportStatus = "pending"    // 待处理
	ReportStatusHandled    ReportStatus = "handled"    // 已处理
	ReportStatusIgnored    ReportStatus = "ignored"    // 已忽略
	ReportStatusOverturned ReportStatus = "overturned" // 申诉成功，撤销原治理决定
)

// ReportAction 管理员对内容采取的动作。warn 只记录并通知，moderated_hidden
// 保留内容供作者整改/申诉，delete 表示永久删除。
const (
	ReportActionWarn            = "warn"
	ReportActionModeratedHidden = "moderated_hidden"
	ReportActionDelete          = "delete"
)

// Report 举报
type Report struct {
	ID             uint         `gorm:"primaryKey" json:"id"`
	ReporterID     uint         `gorm:"not null;index" json:"reporter_id"`
	TargetType     string       `gorm:"size:20;not null;index" json:"target_type"` // post/reply/teacher_rating/major_rating
	TargetID       uint         `gorm:"not null;index" json:"target_id"`
	ReasonCode     string       `gorm:"size:50;index" json:"reason_code"`
	Reason         string       `gorm:"type:text;not null" json:"reason"`
	TargetAuthorID *uint        `gorm:"index" json:"target_author_id"`
	TargetSnapshot string       `gorm:"type:text" json:"target_snapshot"`
	Action            string       `gorm:"size:50" json:"action"`
	ModeratedRevision int          `gorm:"index" json:"moderated_revision,omitempty"`
	Status         ReportStatus `gorm:"default:pending;index" json:"status"`
	HandlerID      *uint        `json:"handler_id"`
	Result         string       `gorm:"size:500" json:"result"`        // 处理结果说明
	DeleteReason   string       `gorm:"size:500" json:"delete_reason"` // 删除理由
	CreatedAt      time.Time    `json:"created_at"`
	HandledAt      *time.Time   `json:"handled_at"`
	Reporter       User         `gorm:"foreignKey:ReporterID" json:"reporter"`
	Handler        *User        `gorm:"foreignKey:HandlerID" json:"handler"`
}

// AppealStatus 申诉状态
type AppealStatus string

const (
	AppealStatusPending AppealStatus = "pending"         // 待投票
	AppealStatusPass    AppealStatus = "pass"            // 申诉成功
	AppealStatusReject  AppealStatus = "reject"          // 申诉失败
	AppealStatusReview  AppealStatus = "review_required" // 平票或法定人数不足，转人工复核
)

// AppealMinRequiredVotes 是结案所需的法定票数下限。
//
// 即时结案（handlers/appeal.go）与到期兜底结案（tasks/appeal_finalizer.go）必须
// 共用同一个阈值，否则改阈值时容易只改一处，出现"投票页说够了、到期结案说不够"。
const AppealMinRequiredVotes = 5

// Appeal 申诉
type Appeal struct {
	ID                   uint         `gorm:"primaryKey" json:"id"`
	ReportID             *uint        `gorm:"uniqueIndex:idx_appeal_report;index" json:"report_id"`
	TargetType           string       `gorm:"size:20;not null;default:post" json:"target_type"`
	TargetID             uint         `gorm:"index" json:"target_id"`
	PostID               uint         `gorm:"not null" json:"post_id"`
	AppellantID          uint         `gorm:"not null" json:"appellant_id"`
	AdminID              uint         `gorm:"not null" json:"admin_id"` // 处理此举报的管理员
	AppellantReason      string       `gorm:"type:text" json:"appellant_reason"`
	EvidenceSnapshot     string       `gorm:"type:text" json:"evidence_snapshot"`
	OriginalPostStatus   PostStatus   `gorm:"size:20" json:"original_post_status"`
	OriginalTargetStatus string       `gorm:"size:20" json:"original_target_status"`
	AdminReason          string       `gorm:"size:500" json:"admin_reason"` // 管理员删除理由
	Status               AppealStatus `gorm:"default:pending" json:"status"`
	Result               string       `gorm:"size:500" json:"result"` // 最终结果
	VotingDeadline       *time.Time   `gorm:"index" json:"voting_deadline"`
	RequiredVotes        int          `gorm:"not null;default:1" json:"required_votes"`
	ClosedReason         string       `gorm:"size:100" json:"closed_reason"`
	EscalationReason     string       `gorm:"size:50" json:"escalation_reason"`
	CreatedAt            time.Time    `json:"created_at"`
	ClosedAt             *time.Time   `json:"closed_at"`
	ReviewedByID         *uint        `json:"reviewed_by_id"`
	ReviewReason         string       `gorm:"size:500" json:"review_reason"`
	ReviewedAt           *time.Time   `json:"reviewed_at"`
	Appellant            User         `gorm:"foreignKey:AppellantID" json:"appellant"`
	Admin                User         `gorm:"foreignKey:AdminID" json:"admin"`
	Post                 Post         `gorm:"foreignKey:PostID" json:"post"`
}

// AppealVote 申诉投票
type AppealVote struct {
	ID           uint      `gorm:"primaryKey" json:"id"`
	AppealID     uint      `gorm:"not null;uniqueIndex:idx_appeal_voter" json:"appeal_id"`
	VoterID      uint      `gorm:"not null;uniqueIndex:idx_appeal_voter" json:"voter_id"`
	Vote         string    `gorm:"size:10;not null" json:"vote"` // support/oppose
	Comment      string    `gorm:"size:500" json:"comment"`
	Recused      bool      `gorm:"default:false" json:"recused"`
	RecuseReason string    `gorm:"size:500" json:"-"`
	CreatedAt    time.Time `json:"created_at"`
	Voter        User      `gorm:"foreignKey:VoterID" json:"voter"`
}

// RectificationReviewStatus 整改复审状态。
type RectificationReviewStatus string

const (
	RectificationReviewPending  RectificationReviewStatus = "pending"
	RectificationReviewApproved RectificationReviewStatus = "approved"
	RectificationReviewRejected RectificationReviewStatus = "rejected"
	RectificationReviewObsolete RectificationReviewStatus = "obsolete"
)

// PostRectificationReview 记录作者提交给管理员复核的明确内容版本。
type PostRectificationReview struct {
	ID                uint                      `gorm:"primaryKey" json:"id"`
	PostID            uint                      `gorm:"not null;index" json:"post_id"`
	ReportID          *uint                     `gorm:"index" json:"report_id,omitempty"`
	SubmittedRevision int                       `gorm:"not null" json:"submitted_revision"`
	Status            RectificationReviewStatus `gorm:"size:20;not null;default:pending;index" json:"status"`
	ReviewerID        *uint                     `gorm:"index" json:"reviewer_id,omitempty"`
	ReviewReason      string                    `gorm:"size:1000" json:"review_reason,omitempty"`
	CreatedAt         time.Time                 `json:"created_at"`
	ReviewedAt        *time.Time                `json:"reviewed_at,omitempty"`
	Post              Post                      `gorm:"foreignKey:PostID" json:"post,omitempty"`
	// Report 是触发治理隐藏的举报记录，仅管理端整改待办用于回溯“为什么当初被处理”。
	// 不进 JSON：Report 含 reporter_id 与举报人自述，不属于审核依据，不能随审核响应外发。
	Report *Report `gorm:"foreignKey:ReportID;-:migration" json:"-"`
}

// PublicAppealUserResponse 是申诉接口允许展示的最小用户资料。
type PublicAppealUserResponse struct {
	ID       uint   `json:"id"`
	Nickname string `json:"nickname"`
	Avatar   string `json:"avatar"`
}

type AppealPostResponse struct {
	ID      uint       `json:"id"`
	Title   string     `json:"title"`
	Content string     `json:"content"`
	Status  PostStatus `json:"status"`
}

type AppealResponse struct {
	ID                   uint                     `json:"id"`
	ReportID             *uint                    `json:"report_id,omitempty"`
	TargetType           string                   `json:"target_type"`
	TargetID             uint                     `json:"target_id"`
	PostID               uint                     `json:"post_id"`
	AppellantReason      string                   `json:"appellant_reason"`
	EvidenceSnapshot     string                   `json:"evidence_snapshot,omitempty"`
	OriginalPostStatus   PostStatus               `json:"original_post_status,omitempty"`
	OriginalTargetStatus string                   `json:"original_target_status,omitempty"`
	AdminReason          string                   `json:"admin_reason"`
	Status               AppealStatus             `json:"status"`
	Result               string                   `json:"result"`
	VotingDeadline       *time.Time               `json:"voting_deadline"`
	RequiredVotes        int                      `json:"required_votes"`
	ClosedReason         string                   `json:"closed_reason"`
	EscalationReason     string                   `json:"escalation_reason,omitempty"`
	CreatedAt            time.Time                `json:"created_at"`
	ClosedAt             *time.Time               `json:"closed_at"`
	ReviewedByID         *uint                    `json:"reviewed_by_id,omitempty"`
	ReviewReason         string                   `json:"review_reason,omitempty"`
	ReviewedAt           *time.Time               `json:"reviewed_at,omitempty"`
	Appellant            PublicAppealUserResponse `json:"appellant"`
	Admin                PublicAppealUserResponse `json:"admin"`
	Post                 AppealPostResponse       `json:"post"`
	CanVote              bool                     `json:"can_vote"`
	CanRecuse            bool                     `json:"can_recuse"`
	IsRecused            bool                     `json:"is_recused"`
	IsAppellant          bool                     `json:"is_appellant"`
	IsAdmin              bool                     `json:"is_admin"`
	MyVote               string                   `json:"my_vote,omitempty"`
	HasVoted             bool                     `json:"has_voted"`
	CastCount            int                      `json:"cast_count,omitempty"`
	SupportCount         int                      `json:"support_count,omitempty"`
	OpposeCount          int                      `json:"oppose_count,omitempty"`
}

type AppealVoteResponse struct {
	ID        uint                     `json:"id"`
	AppealID  uint                     `json:"appeal_id"`
	Vote      string                   `json:"vote"`
	Comment   string                   `json:"comment"`
	Recused   bool                     `json:"recused"`
	CreatedAt time.Time                `json:"created_at"`
	Voter     PublicAppealUserResponse `json:"voter"`
}

// PublicAppealResponse 结案公示 DTO，不包含当事人、管理员或陪审员身份。
type PublicAppealResponse struct {
	ID               uint         `json:"id"`
	PostTitle        string       `json:"post_title"`
	Status           AppealStatus `json:"status"`
	Result           string       `json:"result"`
	ResolutionSource string       `json:"resolution_source"`
	ClosedReason     string       `json:"closed_reason"`
	EscalationReason string       `json:"escalation_reason,omitempty"`
	ClosedAt         *time.Time   `json:"closed_at"`
	CreatedAt        time.Time    `json:"created_at"`
	SupportCount     int          `json:"support_count"`
	OpposeCount      int          `json:"oppose_count"`
}

File: client/lib/models/report.dart
import 'user.dart';

// 公告模型
class Announcement {
  final int id;
  final String title;
  final String content;
  final bool isPinned;
  final int createdBy;
  final User? creator;
  final DateTime createdAt;

  Announcement({
    required this.id,
    required this.title,
    required this.content,
    this.isPinned = false,
    required this.createdBy,
    this.creator,
    required this.createdAt,
  });

  factory Announcement.fromJson(Map<String, dynamic> json) {
    return Announcement(
      id: json['id'] ?? 0,
      title: json['title'] ?? '',
      content: json['content'] ?? '',
      isPinned: json['is_pinned'] ?? false,
      createdBy: json['created_by'] ?? 0,
      creator: json['creator'] != null ? User.fromJson(json['creator']) : null,
      createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
    );
  }
}

// 举报模型
class Report {
  final int id;
  final int reporterId;
  final String targetType;
  final int targetId;
  final String reason;
  final String status;
  final int? handlerId;
  final String result;
  final String deleteReason;
  final DateTime createdAt;
  final DateTime? handledAt;
  final User? reporter;
  final User? handler;

  Report({
    required this.id,
    required this.reporterId,
    required this.targetType,
    required this.targetId,
    required this.reason,
    this.status = 'pending',
    this.handlerId,
    this.result = '',
    this.deleteReason = '',
    required this.createdAt,
    this.handledAt,
    this.reporter,
    this.handler,
  });

  factory Report.fromJson(Map<String, dynamic> json) {
    return Report(
      id: json['id'] ?? 0,
      reporterId: json['reporter_id'] ?? 0,
      targetType: json['target_type'] ?? '',
      targetId: json['target_id'] ?? 0,
      reason: json['reason'] ?? '',
      status: json['status'] ?? 'pending',
      handlerId: json['handler_id'],
      result: json['result'] ?? '',
      deleteReason: json['delete_reason'] ?? '',
      createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
      handledAt: json['handled_at'] != null
          ? DateTime.tryParse(json['handled_at'])
          : null,
      reporter: json['reporter'] != null
          ? User.fromJson(json['reporter'])
          : null,
      handler: json['handler'] != null ? User.fromJson(json['handler']) : null,
    );
  }
}

// 申诉模型
class Appeal {
  final int id;
  final int postId;
  final int appellantId;
  final int adminId;
  final String adminReason;
  final String status;
  final String result;
  final DateTime createdAt;
  final DateTime? closedAt;
  final User? appellant;
  final User? admin;

  Appeal({
    required this.id,
    required this.postId,
    required this.appellantId,
    required this.adminId,
    required this.adminReason,
    this.status = 'pending',
    this.result = '',
    required this.createdAt,
    this.closedAt,
    this.appellant,
    this.admin,
  });

  factory Appeal.fromJson(Map<String, dynamic> json) {
    return Appeal(
      id: json['id'] ?? 0,
      postId: json['post_id'] ?? 0,
      appellantId: json['appellant_id'] ?? 0,
      adminId: json['admin_id'] ?? 0,
      adminReason: json['admin_reason'] ?? '',
      status: json['status'] ?? 'pending',
      result: json['result'] ?? '',
      createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
      closedAt: json['closed_at'] != null
          ? DateTime.tryParse(json['closed_at'])
          : null,
      appellant: json['appellant'] != null
          ? User.fromJson(json['appellant'])
          : null,
      admin: json['admin'] != null ? User.fromJson(json['admin']) : null,
    );
  }
}

// 申诉投票模型
class AppealVote {
  final int id;
  final int appealId;
  final int voterId;
  final String vote;
  final String comment;
  final DateTime createdAt;
  final User? voter;

  AppealVote({
    required this.id,
    required this.appealId,
    required this.voterId,
    this.vote = '',
    this.comment = '',
    required this.createdAt,
    this.voter,
  });

  factory AppealVote.fromJson(Map<String, dynamic> json) {
    return AppealVote(
      id: json['id'] ?? 0,
      appealId: json['appeal_id'] ?? 0,
      voterId: json['voter_id'] ?? 0,
      vote: json['vote'] ?? '',
      comment: json['comment'] ?? '',
      createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
      voter: json['voter'] != null ? User.fromJson(json['voter']) : null,
    );
  }
}

// 邀请模型
class Invitation {
  final int id;
  final int userId;
  final int inviterId;
  final String status;
  final DateTime createdAt;
  final DateTime? acceptedAt;
  final User? user;
  final User? inviter;

  Invitation({
    required this.id,
    required this.userId,
    required this.inviterId,
    this.status = 'pending',
    required this.createdAt,
    this.acceptedAt,
    this.user,
    this.inviter,
  });

  factory Invitation.fromJson(Map<String, dynamic> json) {
    return Invitation(
      id: json['id'] ?? 0,
      userId: json['user_id'] ?? 0,
      inviterId: json['inviter_id'] ?? 0,
      status: json['status'] ?? 'pending',
      createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
      acceptedAt: json['accepted_at'] != null
          ? DateTime.tryParse(json['accepted_at'])
          : null,
      user: json['user'] != null ? User.fromJson(json['user']) : null,
      inviter: json['inviter'] != null ? User.fromJson(json['inviter']) : null,
    );
  }
}
